File: apps/mobile/modules/termix-ssh/android/src/main/java/expo/modules/termixssh/TermixSSHModule.kt
package expo.modules.termixssh

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.functions.Coroutine
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import expo.modules.kotlin.modules.ModuleDefinition
import com.termix.mobilessh.Mobilessh

class TermixSSHModule : Module() {
  private val engine = Mobilessh.newEngine()
  override fun definition() = ModuleDefinition {
    Name("TermixSSH")
    AsyncFunction("start") { config: String -> engine.start(config) }
    AsyncFunction("poll") { id: String -> engine.poll(id) }
    AsyncFunction("trust") { id: String, accept: Boolean -> engine.trust(id, accept) }
    AsyncFunction("write") Coroutine { id: String, data: String -> withContext(Dispatchers.IO) { engine.write(id, data) } }
    AsyncFunction("resize") Coroutine { id: String, cols: Int, rows: Int -> withContext(Dispatchers.IO) { engine.resize(id, cols.toLong(), rows.toLong()) } }
    AsyncFunction("disconnect") { id: String -> engine.disconnect(id) }
    AsyncFunction("awsProfileKey") Coroutine { region: String, profile: String -> withContext(Dispatchers.IO) { Mobilessh.awsProfileKey(region, profile) } }
    AsyncFunction("loginAWS") Coroutine { region: String, credentials: String -> withContext(Dispatchers.IO) { Mobilessh.loginAWS(region, credentials) } }
    AsyncFunction("authorizeEKS") Coroutine { raw: String, credentials: String -> withContext(Dispatchers.IO) { Mobilessh.authorizeEKS(raw, credentials) } }
    AsyncFunction("selectKubeAWSProfile") Coroutine { raw: String, value: String -> withContext(Dispatchers.IO) { Mobilessh.selectKubeAWSProfile(raw, value) } }
    AsyncFunction("listKubeNamespaces") Coroutine { raw: String, value: String -> withContext(Dispatchers.IO) { Mobilessh.listKubeNamespaces(raw, value) } }
    AsyncFunction("inspectKubeContexts") Coroutine { raw: String -> withContext(Dispatchers.IO) { Mobilessh.inspectKubeContexts(raw) } }
    AsyncFunction("selectKubeContext") Coroutine { raw: String, name: String -> withContext(Dispatchers.IO) { Mobilessh.selectKubeContext(raw, name) } }
    AsyncFunction("inspectKubeconfig") Coroutine { raw: String -> withContext(Dispatchers.IO) { Mobilessh.inspectKubeconfig(raw) } }
    AsyncFunction("listKubePods") Coroutine { raw: String, namespace: String -> withContext(Dispatchers.IO) { Mobilessh.listKubePods(raw, namespace) } }
    AsyncFunction("listKubeResources") Coroutine { raw: String, namespace: String, kind: String -> withContext(Dispatchers.IO) { Mobilessh.listKubeResources(raw, namespace, kind) } }
    AsyncFunction("getKubeConfigMap") Coroutine { raw: String, namespace: String, name: String -> withContext(Dispatchers.IO) { Mobilessh.getKubeConfigMap(raw, namespace, name) } }
    AsyncFunction("listKubePodMetrics") Coroutine { raw: String, namespace: String -> withContext(Dispatchers.IO) { Mobilessh.listKubePodMetrics(raw, namespace) } }
    AsyncFunction("getKubePodMetrics") Coroutine { raw: String, namespace: String, name: String -> withContext(Dispatchers.IO) { Mobilessh.getKubePodMetrics(raw, namespace, name) } }
    AsyncFunction("getKubeScale") Coroutine { raw: String, namespace: String, kind: String, name: String -> withContext(Dispatchers.IO) { Mobilessh.getKubeScale(raw, namespace, kind, name) } }
    AsyncFunction("updateKubeScale") Coroutine { raw: String, namespace: String, kind: String, name: String, payload: String -> withContext(Dispatchers.IO) { Mobilessh.updateKubeScale(raw, namespace, kind, name, payload) } }
    AsyncFunction("getKubePodContainers") Coroutine { raw: String, namespace: String, name: String -> withContext(Dispatchers.IO) { Mobilessh.getKubePodContainers(raw, namespace, name) } }
    AsyncFunction("getKubePodLogs") Coroutine { raw: String, namespace: String, name: String, container: String, previous: Boolean -> withContext(Dispatchers.IO) { Mobilessh.getKubePodLogs(raw, namespace, name, container, previous) } }
    OnActivityEntersBackground { engine.close() }
    OnDestroy { engine.close() }
  }
}

File: apps/mobile/modules/termix-ssh/ios/TermixSSHModule.swift
import ExpoModulesCore
import TermixSSH

public class TermixSSHModule: Module {
  private let engine = MobilesshNewEngine()!

  public func definition() -> ModuleDefinition {
    Name("TermixSSH")
    Function("mobileSyncCapability") { MobileCloud.capability() }
    AsyncFunction("fetchMobileSettings") { () async -> [String: String] in await MobileCloud.fetch() }
    AsyncFunction("start") { (config: String) in try self.engine.start(config) }
    AsyncFunction("poll") { (id: String) in self.engine.poll(id) }
    AsyncFunction("trust") { (id: String, accept: Bool) in try self.engine.trust(id, accept: accept) }
    AsyncFunction("write") { (id: String, data: String) in try self.engine.write(id, text: data) }.runOnQueue(.global(qos: .userInitiated))
    AsyncFunction("resize") { (id: String, cols: Int, rows: Int) in try self.engine.resize(id, cols: cols, rows: rows) }.runOnQueue(.global(qos: .userInitiated))
    AsyncFunction("disconnect") { (id: String) in self.engine.disconnect(id) }
    AsyncFunction("awsProfileKey") { (region: String, profile: String) -> String in
      var error: NSError?
      let result = MobilesshAWSProfileKey(region, profile, &error)
      if let error { throw error }
      return result
    }.runOnQueue(.global(qos: .userInitiated))
    AsyncFunction("loginAWS") { (region: String, credentials: String) -> String in
      var error: NSError?
      let result = MobilesshLoginAWS(region, credentials, &error)
      if let error { throw error }
      return result
    }.runOnQueue(.global(qos: .userInitiated))
    AsyncFunction("authorizeEKS") { (raw: String, credentials: String) -> String in
      var error: NSError?
      let result = MobilesshAuthorizeEKS(raw, credentials, &error)
      if let error { throw error }
      return result
    }.runOnQueue(.global(qos: .userInitiated))
    AsyncFunction("selectKubeAWSProfile") { (raw: String, value: String) -> String in
      var error: NSError?
      let result = MobilesshSelectKubeAWSProfile(raw, value, &error)
      if let error { throw error }
      return result
    }.runOnQueue(.global(qos: .userInitiated))
    AsyncFunction("listKubeNamespaces") { (raw: String, value: String) -> String in
      var error: NSError?
      let result = MobilesshListKubeNamespaces(raw, value, &error)
      if let error { throw error }
      return result
    }.runOnQueue(.global(qos: .userInitiated))
    AsyncFunction("inspectKubeContexts") { (raw: String) -> String in
      var error: NSError?
      let result = MobilesshInspectKubeContexts(raw, &error)
      if let error { throw error }
      return result
    }.runOnQueue(.global(qos: .userInitiated))
    AsyncFunction("selectKubeContext") { (raw: String, name: String) -> String in
      var error: NSError?
      let result = MobilesshSelectKubeContext(raw, name, &error)
      if let error { throw error }
      return result
    }.runOnQueue(.global(qos: .userInitiated))
    AsyncFunction("inspectKubeconfig") { (raw: String) -> String in
      var error: NSError?
      let result = MobilesshInspectKubeconfig(raw, &error)
      if let error { throw error }
      return result
    }.runOnQueue(.global(qos: .userInitiated))
    AsyncFunction("listKubePods") { (raw: String, namespace: String) -> String in
      var error: NSError?
      let result = MobilesshListKubePods(raw, namespace, &error)
      if let error { throw error }
      return result
    }.runOnQueue(.global(qos: .userInitiated))
    AsyncFunction("listKubeResources") { (raw: String, namespace: String, kind: String) -> String in
      var error: NSError?
      let result = MobilesshListKubeResources(raw, namespace, kind, &error)
      if let error { throw error }
      return result
    }.runOnQueue(.global(qos: .userInitiated))
    AsyncFunction("getKubeConfigMap") { (raw: String, namespace: String, name: String) -> String in
      var error: NSError?
      let result = MobilesshGetKubeConfigMap(raw, namespace, name, &error)
      if let error { throw error }
      return result
    }.runOnQueue(.global(qos: .userInitiated))
    AsyncFunction("listKubePodMetrics") { (raw: String, namespace: String) -> String in
      var error: NSError?
      let result = MobilesshListKubePodMetrics(raw, namespace, &error)
      if let error { throw error }
      return result
    }.runOnQueue(.global(qos: .userInitiated))
    AsyncFunction("getKubePodMetrics") { (raw: String, namespace: String, name: String) -> String in
      var error: NSError?
      let result = MobilesshGetKubePodMetrics(raw, namespace, name, &error)
      if let error { throw error }
      return result
    }.runOnQueue(.global(qos: .userInitiated))
    AsyncFunction("getKubeScale") { (raw: String, namespace: String, kind: String, name: String) -> String in
      var error: NSError?
      let result = MobilesshGetKubeScale(raw, namespace, kind, name, &error)
      if let error { throw error }
      return result
    }.runOnQueue(.global(qos: .userInitiated))
    AsyncFunction("updateKubeScale") { (raw: String, namespace: String, kind: String, name: String, payload: String) -> String in
      var error: NSError?
      let result = MobilesshUpdateKubeScale(raw, namespace, kind, name, payload, &error)
      if let error { throw error }
      return result
    }.runOnQueue(.global(qos: .userInitiated))
    AsyncFunction("getKubePodContainers") { (raw: String, namespace: String, name: String) -> String in
      var error: NSError?
      let result = MobilesshGetKubePodContainers(raw, namespace, name, &error)
      if let error { throw error }
      return result
    }.runOnQueue(.global(qos: .userInitiated))
    AsyncFunction("getKubePodLogs") { (raw: String, namespace: String, name: String, container: String, previous: Bool) -> String in
      var error: NSError?
      let result = MobilesshGetKubePodLogs(raw, namespace, name, container, previous, &error)
      if let error { throw error }
      return result
    }.runOnQueue(.global(qos: .userInitiated))
    AsyncFunction("clearSensitiveImportCopy") { (name: String) in
      // UIKit 匯入模式會另建 Inbox 副本；僅清理本 App 暫存區的同名檔。
      let manager = FileManager.default
      guard !name.isEmpty, name != ".", name != "..",
        !name.contains("/"), !name.contains("\\"),
        let bundleID = Bundle.main.bundleIdentifier else {
        throw NSError(domain: "TermixImport", code: 1)
      }
      let inbox = manager.temporaryDirectory.appendingPathComponent(bundleID + "-Inbox", isDirectory: true)
      let file = inbox.appendingPathComponent(name)
      guard manager.fileExists(atPath: file.path) else { return }
      guard file.resolvingSymlinksInPath().deletingLastPathComponent() == inbox.standardizedFileURL.resolvingSymlinksInPath() else {
        throw NSError(domain: "TermixImport", code: 1)
      }
      let values = try file.resourceValues(forKeys: [.isRegularFileKey, .isSymbolicLinkKey])
      guard values.isRegularFile == true, values.isSymbolicLink != true else {
        throw NSError(domain: "TermixImport", code: 1)
      }
      do { try manager.removeItem(at: file) }
      catch { throw NSError(domain: "TermixImport", code: 1) }
    }.runOnQueue(.global(qos: .userInitiated))
    OnAppEntersBackground { self.engine.close() }
    OnDestroy { self.engine.close() }
  }
}
